### ISSUE TITLE
Removed locked parameters are still valid in embedding

### ISSUE THREAD

**Issue by nakazavva** [metabase/metabase#4779]
- Metabase version: 0.23.1
- Metabase hosting environment: Docker
- Metabase internal database: H2 (default)

In the following step, removed locked parameters are still valid in embedding and the API return 400 "Bad Request", saying "You must specify a value for :xxx in the JWT."

1.  in Sharing > Application, set some parameters with locked.
2. then back to dashboard and remove the filters for that parameters above.

**Comment by nakazavva** [comment]
that is a kind of trivial stuff and i can resolve the following step though.

1. Create again a filter with exactly same name.
2. You can see removed locked parameter is back in embedding parameter section.

just for someone who gets stuck in this issue.

**Comment by camsaul** [comment]
This actually seems like it's two separate issues:

1.  Changing the parameters in a Dashboard will leave the Dashboard's params out of sync with its `embedding_params` (the map of what params are allowed/disabled/locked). 
2.  Editing the `embedding_params` for a Dashboard (ostensibly to fix the parameter mismatch in part 1) doesn't work because of what seems to be a race condition in the frontend code. 

### The first issue

> Changing the parameters in a Dashboard will leave the Dashboard's params out of sync with its `embedding_params` (the map of what params are allowed/disabled/locked). 

What should we do when someone changes the parameters a dashboard accepts when it's embedded? Should we just warn people that changing the parameters for a dashboard that is embedded means they'll need to configure the dash's parameters again and generate new embedding tokens?

Background: there has to be an exact match between the parameters passed in in the embedding token and what's enabled/disabled/locked for the dashboard. Specifying invalid params for a dashboard isn't allowed. 

If you remove a parameter from a dashboard that's embedded that needs to be taken out of the `embedding_params` list as well. But if we did that automatically, any existing embeds will stop working immediately (since they're now passing in an invalid param). Alternatively, we could leave `embedding_params` as-is until someone updates the `embedding_params` in the dashboard embedding config page. (That's what we currently do.) Unfortunately that doesn't work as well as it should because of the second issue. 

Automatically updating `embedding_params` would have a few other problems: 

1.  We don't know what people would want us to do with *new* params (should they be locked, enabled, or disabled?). Since it would immediately affect any existing embeds, assuming the wrong thing could have nasty consequences.
2.  It's tricky to figure out that a param has simply been renamed. Updating the `embedding_params` to reflect a named param would still affect any existing embeds since the params in their token would no longer match up.

FWIW I think warning people about the consequences and possibly requiring them to update the dashboard's embedding params configuration is the safest option. Trying to automatically update things could at best break a few existing embeds and at worst expose data that someone never would have wanted exposed.

But if we were to do something like require someone to reconfigure the embedding params that still wouldn't work, because of issue 2:

### The second issue

> Editing the `embedding_params` for a Dashboard (ostensibly to fix the parameter mismatch in part 1) doesn't work because the old params are still saved in `embedding_params`

This is best explained with steps to repro.

1.  Create a dashboard with a filter `num_toucans`
2.  Go to embedding and embed the dashboard, make the filter "locked". (`embedding_params` is now `{:num_toucans :locked}`)
3.  Go back to the dashboard and change the name of the filter to `toucans`.
4.  Go back to embedding. The preview won't work because `num_toucans` is required.
5.  Select a value for the `toucans` param. Make it locked. Click the button to "Publish" your changes.
6. You would expect it to work at this point, but it doesn't: there's still an error message about `num_toucans` being required. The actual value of `embedding_params` is now `{:num_toucans :locked, :toucans :locked}` (the old `embedding_params` combined with the new ones).

It looks like the issue is that the dashboard embedding page is combining whatever the old value of `embedding_params` is with the new one rather than just replacing the value with the new one.

Another thing I've noticed is what appears to be a race condition: the frontend simultaneously issues requests to update the dashboard (`PUT /api/dashboard/:id`) and preview it (`GET /api/preview_embed/dashboard/:token`). It appears that the preview embed endpoint runs before the new `embedding_params` value is in the DB, meaning it still uses the old value; this can lead to things failing where they'd otherwise work. We should probably update the value and only call the preview endpoint after the update is completed.

**Comment by camsaul** [comment]
@salsakran @mazameli @kdoh the first issue sounds like a UX thing that needs to be thought through

**Comment by shrike71** [comment]
Any update on this issue? It's a really nasty one for us. At this point i'd be happy enough if somebody was able to furnish a workaround to bulk-clear the parameter/s! This is still present in version 0.27.2 .

**Comment by coderberry** [comment]
I am still having this issue as well on v0.29.3

**Comment by jeremyjh** [comment]
Even if you do not fix this, you need to remove the embedded params when you un-publish the dashboard. 

I created a dashboard with a parameter name that apparently is unusable, and now it is impossible for me to fix it in the UI even though I've renamed it, unpublished, republished etc.

**Comment by stevewritescode** [comment]
Still an issue in v0.32.4.

We got stuck for a while trying to figure out why our embedded dashboard was broken before finding out this was the issue after digging into the `reporting_dashboard` table in the backend DB.

For anyone else searching github, this is the error we saw in the logs that eventually led us back to the bug:

`You must specify a value for :category in the JWT`

where `category` happened to be the name of our old parameter that got "stuck" in the DB `embedding_params` record.

**Comment by salsakran** [comment]
bumping this in priority due to a number of folks hitting it.

**Comment by pjatx** [comment]
+1 on this - have a ton of embedded charts and we changed the lookup based on one locked param to another and broke everything. Now we'll have to go through and implement this workaround. 

Think it the expected behavior would be for the configured params in the UI to be the source of truth. IE if I remove a locked param and add a new one, it should then require the new one and no longer the previous as it does today.

**Comment by doc82** [comment]
I'm stuck with this issue. I've removed all my params. Unpublished, and re-published. and I'm getting an error that my JWT is missing a param that no longer exists :(

Any ideas?

**Comment by pjatx** [comment]
@doc82 So what you have to do is add the original field back (exactly the same name) and leave it in there as a placeholder. I just pass through a bogus value and don't use it in the actual query and seems to work fine. The only other option is to create a new question from scratch.

**Comment by flyingmachine** [comment]
## Issue breakdown

Here's my understanding of the issue: Modifying the dashboard's `parameters` attribute does not update the `embedding_params` attribute. If you change a parameter's name from `ricky` to `bobby`, embedding params will still contain some value like `{"ricky": "locked"}`. Likewise, if you delete the parameter, it will remain in `embedding_params`.

I don't think that there's a race condition involved when modifying `embedding_params`. If you load a dashboard with these attributes (clojure):

```clojure
{:parameters       [{:name "ID"
                     :slug "id"
                     :type "id"
                     :id   "uniq-id"}]
 :embedding_params {:id "locked"}}
```

And rename the `"ID"` parameter to `"film ID"`, the resulting dashboard will have these attributes:

```clojure
{:parameters       [{:name "film ID"
                     :slug "film_id"
                     :type "id"
                     :id   "uniq-id"}]
 :embedding_params {:id "locked"}}
```

If you then navigate to the embedding interface, you can set an embedding value for the parameter with the slug `"film_id"`. There is no option to modify the key `:id` in `embedding_params` because there are no longer any parameters with a slug of `"id"`. I don't think it's a race condition. The result will look like this:

```clojure
{:parameters       [{:name "film ID"
                     :slug "film_id"
                     :type "id"
                     :id   "uniq-id"}]
 :embedding_params {:id "locked", :film_id "locked"}}
```

## Proposed solution

My proposal is to store parameters like this:

```clojure
{:parameters [{:name      "film ID"
               :slug      "film_id"
               :id        "570ce143"
               :type      "id"
               :embedding "disabled | editable | locked"}]}
```

To generate a token, users will have something like this on their server:

```clojure
;; this is used to generate JWT token:
(def payload
  {:resource {:dashboard 451}
   :params   {:name "film ID"  ;; this is only used as documentation to help you remember what this refers to
              :id   "570ce143" ;; use the parameter's id, not the slug, for identification
              }
   :exp      "exp"})
```

This will solve the following problems:

* dashboards no longer working if a param gets renamed
* edit preview being broken because of "hidden" locked param that the embedding UI won't show
* in general, we will not have the `parameters` / `embedding_params` sync problem

It won't solve the problem that dashboards will no longer work if you delete a locked parameter. I'm not sure that's actually a problem, though? I think that's the intended behavior.

## Possible problems

This would require everyone to update their servers' JWT token creation code. We could address this issue by continuing to allow JWT token generation using the current process until some future release, giving users X amount of time to upgrade.

**Comment by flyingmachine** [comment]
resolved by #10878
